### PR TITLE
test: add batch size threshold tuning tests for memory extractor

### PR DESCRIPTION
**Pull request by mkoushni**
# Batch Size Threshold Tuning Tests

## Summary

This PR adds comprehensive test coverage for batch size threshold functionality in the memory extractor. The tests validate that memory extraction runs every N turns (default: 10) to avoid excessive LLM calls, ensuring proper turn count tracking, batch selection logic, and boundary condition handling.

### Files Modified

- `src/semantic-router/pkg/memory/extractor_test.go` - Added 13 new test functions 

### New Tests Added (13 functions, 18 test cases)

1. **`TestProcessResponse_DefaultBatchSize`**
   - Tests default batch size of 10 turns
   - Verifies extraction only happens when `turnCount % 10 == 0`
   - Tests turns 1-9 (no extraction) and turn 10 (extraction)

2. **`TestProcessResponse_DefaultBatchSizeBoundary`**
   - Tests boundary conditions at exactly turn 10
   - Verifies turn 9 does NOT extract (9 % 10 != 0)
   - Verifies turn 10 DOES extract (10 % 10 == 0)
   - Verifies turn 11 does NOT extract again (11 % 10 != 0)

3. **`TestProcessResponse_CustomBatchSize`**
   - Tests custom batch size of 5 turns
   - Verifies extraction happens at turn 5 (5 % 5 == 0)
   - Validates turns 1-4 do not extract

4. **`TestProcessResponse_BatchSizeZeroUsesDefault`**
   - Tests that BatchSize = 0 defaults to 10
   - Verifies default behavior when batch size is not configured

5. **`TestProcessResponse_BatchSizeNegativeUsesDefault`**
   - Tests that BatchSize < 0 defaults to 10
   - Validates error handling for invalid batch size values

6. **`TestProcessResponse_MultipleSessionsIndependent`**
   - Tests that turn counts are tracked independently per session
   - Verifies session isolation (session1 and session2 have separate turn counts)
   - Ensures extraction happens per-session based on individual turn counts

7. **`TestProcessResponse_BatchSelectionLogic`**
   - Tests batch selection logic (last N+5 messages for context)
   - Verifies that when history > batchSize+5, only last N+5 messages are used
   - Validates batch extraction includes sufficient context

8. **`TestProcessResponse_BatchSelectionSmallHistory`**
   - Tests batch selection with history smaller than batchSize+5
   - Verifies all messages are included when history is small
   - Ensures extraction works correctly with limited history

9. **`TestProcessResponse_TurnCountTracking`**
   - Tests turn count tracking across multiple turns
   - Verifies extraction happens at correct intervals (e.g., turns 3, 6 for batch size 3)
   - Validates turn count increments correctly per session

10. **`TestProcessResponse_StoreDisabled`**
    - Tests behavior when store is disabled
    - Verifies ProcessResponse returns without error when store is not enabled
    - Ensures graceful handling of disabled store

11. **`TestProcessResponse_ExtractionDisabled`**
    - Tests behavior when extraction is disabled in config
    - Verifies ProcessResponse returns without error when extraction is disabled
    - Ensures graceful handling of disabled extraction

12. **`TestProcessResponse_EmptyEndpoint`**
    - Tests behavior when endpoint is empty
    - Verifies ProcessResponse returns without error when endpoint is not configured
    - Ensures graceful handling of missing endpoint

13. **`TestProcessResponse_MultipleBatchSizes`**
    - Comprehensive test with multiple batch sizes (1, 3, 5, 10, 20)
    - Validates extraction happens at correct turns for each batch size
    - Tests boundary conditions for various batch size configurations
    - 5 subtests covering different batch sizes

### Test Results

All tests pass successfully:
```
✅ TestProcessResponse_DefaultBatchSize
✅ TestProcessResponse_DefaultBatchSizeBoundary
✅ TestProcessResponse_CustomBatchSize
✅ TestProcessResponse_BatchSizeZeroUsesDefault
✅ TestProcessResponse_BatchSizeNegativeUsesDefault
✅ TestProcessResponse_MultipleSessionsIndependent
✅ TestProcessResponse_BatchSelectionLogic
✅ TestProcessResponse_BatchSelectionSmallHistory
✅ TestProcessResponse_TurnCountTracking
✅ TestProcessResponse_StoreDisabled
✅ TestProcessResponse_ExtractionDisabled
✅ TestProcessResponse_EmptyEndpoint
✅ TestProcessResponse_MultipleBatchSizes (5 subtests)
```

### 1. Default Batch Size (10 turns)
- ✅ Verification of default batch size: 10 turns
- ✅ Extraction only at turns divisible by 10 (10, 20, 30, ...)
- ✅ No extraction at other turns (1-9, 11-19, ...)

### 2. Custom Batch Sizes
- ✅ Custom batch size configurations (1, 3, 5, 20)
- ✅ Extraction at correct intervals for each batch size
- ✅ Validation of batch size parameter

### 3. Boundary Conditions
- ✅ Exact boundary at batch size (turn 10 for batch size 10)
- ✅ Just before boundary (turn 9 for batch size 10)
- ✅ Just after boundary (turn 11 for batch size 10)
- ✅ Multiple boundaries (turns 3, 6, 9 for batch size 3)

### 4. Turn Count Tracking
- ✅ Per-session turn count tracking
- ✅ Independent turn counts for different sessions
- ✅ Turn count increments correctly
- ✅ Extraction based on session-specific turn count

### 5. Batch Selection Logic
- ✅ Last N+5 messages selected for context
- ✅ Full history used when history < batchSize+5
- ✅ Correct message range selected for extraction

### 6. Edge Cases
- ✅ BatchSize = 0 (defaults to 10)
- ✅ BatchSize < 0 (defaults to 10)
- ✅ Disabled store (graceful handling)
- ✅ Disabled extraction (graceful handling)
- ✅ Empty endpoint (graceful handling)
- ✅ Small history (all messages included)

### Batch Size Logic

The memory extractor uses batch size to control when extraction runs:

1. **Default Batch Size: 10 turns**
   - Extraction runs every 10 turns by default
   - Configurable via `ExtractionConfig.BatchSize`
   - If BatchSize <= 0, defaults to 10

2. **Turn Count Tracking**
   - Turn counts tracked per `sessionID`
   - Each call to `ProcessResponse` increments turn count for that session
   - Extraction happens when `turnCount % batchSize == 0`

3. **Batch Selection**
   - When extraction runs, selects last `batchSize + 5` messages for context
   - If history < batchSize + 5, uses all available messages
   - Provides sufficient context for LLM extraction

4. **Extraction Trigger**
   ```go
   if turnCount % batchSize != 0 {
       return nil  // Skip extraction
   }
   // Extract from batch
   ```

### Test Approach

- Uses mock LLM server (`httptest.Server`) for isolated testing
- Tests with `InMemoryStore` for fast, isolated storage testing
- Validates both extraction triggering and batch selection
- Tests cover all code paths in `ProcessResponse()`
- Includes edge cases and error conditions

## Checklist

- [x] All new tests pass
- [x] Existing tests still pass
- [x] Code coverage verified
- [x] Boundary conditions tested
- [x] Custom batch sizes tested
- [x] Turn count tracking validated
- [x] Session isolation verified
- [x] Batch selection logic tested
- [x] Edge cases covered
- [x] No breaking changes
- [x] Tests are well-documented

Resolve https://github.com/yehudit1987/semantic-router/issues/11

### CODE REVIEW

File: src/semantic-router/pkg/memory/extractor_test.go [hunk 1]
```diff
@@ -490,10 +494,413 @@ func TestTruncateForLog(t *testing.T) {
...
+	assert.Equal(t, int32(2), atomic.LoadInt32(callCount), "session2 should extract at turn 5")
+}
+
+// TestProcessResponse_BatchSelectionLogic tests that when history > batchSize+5,
+// only the last batchSize+5 messages are used for extraction.
+func TestProcessResponse_BatchSelectionLogic(t *testing.T) {
+	var mu sync.Mutex
+	var capturedBody []byte
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		defer r.Body.Close()
```
Comment (yehuditkerido) [inline src/semantic-router/pkg/memory/extractor_test.go:650]: **Minor:** this HTTP handler runs in a separate goroutine, and `require.NoError` is intended to be called only from the test goroutine. If an error occurs, it could terminate the test run instead of failing the test normally. The risk here is low, but it may be worth replacing it with `assert.NoError` or passing the error back to the test goroutine for checking, to improve test stability.

File: src/semantic-router/pkg/memory/extractor_test.go [hunk 2]
```diff
@@ -490,10 +494,413 @@ func TestTruncateForLog(t *testing.T) {
...
+	assert.Contains(t, bodyStr, "message_18", "message_18 should be in batch")
+}
+
+// TestProcessResponse_BatchSelectionSmallHistory tests batch selection with history
+// smaller than batchSize+5. Verifies all messages are included.
+func TestProcessResponse_BatchSelectionSmallHistory(t *testing.T) {
+	var mu sync.Mutex
+	var capturedBody []byte
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		defer r.Body.Close()
```
Comment (yehuditkerido) [inline src/semantic-router/pkg/memory/extractor_test.go:709]: Same minor issue as in `TestProcessResponse_BatchSelectionLogic`.

**Review by Xunzhuo — APPROVED** [review]
